### PR TITLE
All Remaining Command Cleanup

### PR DESCRIPTION
**Pull request by openbots-ff**
Cleanup:
- DataBase Commands
- Engine Commands
- Input Commands
- Misc Commands
- NLG Commands
- Regex Commands
- Remote Commands
- System Commands
- Window Commands

Added new CommandsControls:
- CreateDefaultPasswordInputGroupFor - Hides text typed into TextBoxes by replacing it with '*'. Used primarily for password inputs. 
- CreateDefaultWindowControlGroupFor - Dropdown that generates all currently open windows. Used frequently in window and input commands. 

Fixed a few leftover variable clarity bugs. Specifically in the variable manager when editing variables/elements.
Added web automation engine options for microsoft edge and firefox. 
Unused Nuget packages have been removed from their respective projects.

### CODE REVIEW

File: taskt.Commands/Data Commands/DateCalculationCommand.cs [hunk 1]
```diff
@@ -163,32 +198,22 @@ public override List<Control> Render(IfrmCommandEditor editor)
...
+            if (operand == "Add")
+                operandLanguage = "to";
             else
-            {
-                return base.GetDisplayValue();
-            }
+                operandLanguage = "from";
+
+            if (operand == "Get")
+                operand = v_CalculationMethod.Replace(interval, "").TrimEnd();//$"{v_CalculationMethod.Split(' ')[0]} {v_CalculationMethod.Split(' ')[1]}";
```
Comment (openbots-ab) [inline taskt.Commands/Data Commands/DateCalculationCommand.cs:213]: Please remove the comment.

File: taskt.Commands/Database Commands/DefineDatabaseConnectionCommand.cs [hunk 2]
```diff
@@ -33,17 +31,17 @@ public class DefineDatabaseConnectionCommand : ScriptCommand
         public string v_InstanceName { get; set; }
 
         [XmlAttribute]
-        [PropertyDescription("Define Connection String")]
-        [InputSpecification("")]
-        [SampleUsage("")]
+        [PropertyDescription("Connection String")]
+        [InputSpecification("Define the string to use when connecting to the OleDb database.")]
+        [SampleUsage("myRobot || {vUsername}")]
```
Comment (openbots-ab) [inline taskt.Commands/Database Commands/DefineDatabaseConnectionCommand.cs:36]:
```suggestion
        [SampleUsage("Provider=sqloledb;Data Source=myServerAddress;Initial Catalog=myDataBase;Integrated Security=SSPI; || {vConnectionString}")]
```

File: taskt.Commands/Regex Commands/RegexIsMatchCommand.cs [hunk 3]
```diff
@@ -17,26 +17,24 @@ namespace taskt.Commands
...
+        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
+        public string v_InputText { get; set; }
 
         [XmlAttribute]
-        [PropertyDescription("Please enter regex pattern")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
-        [InputSpecification("Enter a Regex Pattern to apply to check if a match exists")]
-        [SampleUsage(@"**^([\w\-]+)** or **vSomeVariable**")]
+        [PropertyDescription("Regex Pattern")]
+        [InputSpecification("Enter a Regex Pattern to apply to Input Text to get all the matches.")]
```
Comment (openbots-ab) [inline taskt.Commands/Regex Commands/RegexIsMatchCommand.cs:33]:
```suggestion
        [InputSpecification("Enter a Regex Pattern to apply to the Input Text.")]
```

File: taskt.Commands/Regex Commands/RegexReplaceCommand.cs [hunk 4]
```diff
@@ -17,34 +17,33 @@ namespace taskt.Commands
 {
     [Serializable]
     [Group("Regex Commands")]
-    [Description("This command allows you to replace all the matches in Text based on RegEx")]
-    [UsesDescription("Use this command when you want to replace all matches in text based on Regex Pattern")]
-    [ImplementationDescription("This command implements Replace Action of Regex for given input Text and Regex Pattern and returns a text after replacement")]
+    [Description("This command replaces all the matches in a given text based on a Regex pattern")]
```
Comment (openbots-ab) [inline taskt.Commands/Regex Commands/RegexReplaceCommand.cs:20]:
```suggestion
    [Description("This command replaces all the matches in a given text based on a Regex pattern.")]
```

File: taskt.Commands/Regex Commands/RegexReplaceCommand.cs [hunk 5]
```diff
@@ -17,34 +17,33 @@ namespace taskt.Commands
...
+        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
+        public string v_InputText { get; set; }
 
         [XmlAttribute]
-        [PropertyDescription("Please enter regex pattern")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
-        [InputSpecification("Enter a Regex Pattern to apply to replace matches with given text")]
-        [SampleUsage(@"**^([\w\-]+)** or **vSomeVariable**")]
+        [PropertyDescription("Regex Pattern")]
+        [InputSpecification("Enter a Regex Pattern to apply to Input Text to get all the matches.")]
```
Comment (openbots-ab) [inline taskt.Commands/Regex Commands/RegexReplaceCommand.cs:33]:
```suggestion
        [InputSpecification("Enter a Regex Pattern to apply to the Input Text.")]
```

File: taskt.Commands/Regex Commands/RegexSplitCommand.cs [hunk 6]
```diff
@@ -18,29 +18,27 @@ namespace taskt.Commands
...
+        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
+        public string v_InputText { get; set; }
 
         [XmlAttribute]
-        [PropertyDescription("Please enter regex pattern")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
-        [InputSpecification("Enter a Regex Pattern to apply to split text based on RegEx")]
-        [SampleUsage(@"**^([\w\-]+)** or **vSomeVariable**")]
+        [PropertyDescription("Regex Pattern")]
+        [InputSpecification("Enter a Regex Pattern to apply to Input Text to get all the matches.")]
```
Comment (openbots-ab) [inline taskt.Commands/Regex Commands/RegexSplitCommand.cs:34]:
```suggestion
        [InputSpecification("Enter a Regex Pattern to apply to the Input Text.")]
```

File: taskt.Commands/Remote Commands/RemoteAPICommand.cs [hunk 7]
```diff
@@ -14,36 +14,34 @@
...
         public string v_BaseURL { get; set; }
 
         [XmlAttribute]
-        [PropertyDescription("Select Parameter Type")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
+        [PropertyDescription("Parameter Type")]
         [PropertyUISelectionOption("Get Engine Status")]
         [PropertyUISelectionOption("Restart taskt")]
-        [InputSpecification("Select the necessary API Method")]
+        [InputSpecification("Select the appropriate parameter.")]
```
Comment (openbots-ab) [inline taskt.Commands/Remote Commands/RemoteAPICommand.cs:34]:
```suggestion
        [InputSpecification("Select the appropriate parameter type.")]
```

File: taskt.Commands/Remote Commands/RemoteTaskCommand.cs [hunk 8]
```diff
@@ -17,52 +17,52 @@ namespace taskt.Commands
...
         [XmlAttribute]
-        [PropertyDescription("Select Parameter Type")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
+        [PropertyDescription("Parameter Type")]
         [PropertyUISelectionOption("Run Raw Script Data")]
         [PropertyUISelectionOption("Run Local File")]
         [PropertyUISelectionOption("Run Remote File")]
         [PropertyUISelectionOption("Run Command Json")]
-        [InputSpecification("Select the necessary parameter.")]
+        [InputSpecification("Select the appropriate parameter.")]
```
Comment (openbots-ab) [inline taskt.Commands/Remote Commands/RemoteTaskCommand.cs:37]:
```suggestion
        [InputSpecification("Select the appropriate parameter type.")]
```

File: taskt.Commands/System Commands/EnvironmentVariableCommand.cs [hunk 9]
```diff
@@ -17,15 +17,12 @@ namespace taskt.Commands
 {
     [Serializable]
     [Group("System Commands")]
-    [Description("This command allows you to exclusively select a system/environment variable")]
-    [UsesDescription("Use this command to exclusively retrieve a system variable")]
-    [ImplementationDescription("")]
+    [Description("This command exclusively selects a environment variable.")]
```
Comment (openbots-ab) [inline taskt.Commands/System Commands/EnvironmentVariableCommand.cs:20]:
```suggestion
    [Description("This command exclusively selects an environment variable.")]
```

File: taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs [hunk 10]
```diff
@@ -16,49 +16,47 @@ namespace taskt.Commands
 {
     [Serializable]
     [Group("System Commands")]
-    [Description("This command allows you to stop a program or a process.")]
-    [UsesDescription("Use this command to close an application by its name such as 'chrome'. Alternatively, you may use the Close Window or Thick App Command instead.")]
-    [ImplementationDescription("This command implements 'Process.CloseMainWindow'.")]
+    [Description("This command allows you to launch a remote desktop session.")]
```
Comment (openbots-ab) [inline taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs:19]:
```suggestion
    [Description("This command launches a remote desktop session.")]
```

File: taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs [hunk 11]
```diff
@@ -16,49 +16,47 @@ namespace taskt.Commands
...
     public class LaunchRemoteDesktopCommand : ScriptCommand
     {
         [XmlAttribute]
-        [PropertyDescription("Enter the name of the machine")]
-        [PropertyUIHelper(UIAdditionalHelperType.ShowVariableHelper)]
-        [InputSpecification("")]
+        [PropertyDescription("Machine Name")]
+        [InputSpecification("Define the name of the machine to log on to.")]
         [SampleUsage("")]
```
Comment (openbots-ab) [inline taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs:25]: Please provide Sample Usage.

File: taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs [hunk 12]
```diff
@@ -5,45 +5,39 @@
...
 using taskt.UI.CustomControls;
 
 namespace taskt.Commands
 {
     [Serializable]
     [Group("Engine Commands")]
-    [Description("This command allows you to set preferences for engine behavior.")]
-    [UsesDescription("Use this command when you want to change the engine behavior.")]
-    [ImplementationDescription("")]
+    [Description("This command allows you to set preferences for engine behavior at runtime.")]
```
Comment (openbots-ab) [inline taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs:16]:
```suggestion
    [Description("This command sets preferences for engine behavior at runtime.")]
```

File: taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs [hunk 13]
```diff
@@ -52,22 +46,23 @@ public override void RunCommand(object sender)
...
-            RenderedControls.AddRange(CommandControls.CreateDefaultDropdownGroupFor("v_PreferenceType", this, editor));
+            RenderedControls.AddRange(CommandControls.CreateDefaultDropdownGroupFor("v_CalculationOption", this, editor));
 
             return RenderedControls;
         }
+
         public override string GetDisplayValue()
         {
-            return base.GetDisplayValue() + $" [{v_PreferenceType}]";
+            return base.GetDisplayValue() + $" [Calculation '{v_CalculationOption}']";
```
Comment (openbots-ab) [inline taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs:65]: I think it should simply be
```suggestion
            return base.GetDisplayValue() + $" [To '{v_CalculationOption}']";
```

File: taskt.Studio/Commands/Window Commands/SetWindowStateCommand.cs [hunk 14]
```diff
@@ -83,33 +80,21 @@ public override void RunCommand(object sender)
...
-            RenderedControls.Add(WindowNameControl);
-
-            var windowStateLabel = CommandControls.CreateDefaultLabelFor("v_WindowState", this);
-            RenderedControls.Add(windowStateLabel);
-
-            var windowStateControl = CommandControls.CreateDropdownFor("v_WindowState", this);
-            RenderedControls.Add(windowStateControl);
+            RenderedControls.AddRange(CommandControls.CreateDefaultWindowControlGroupFor("v_WindowName", this, editor));
+            RenderedControls.Add(CommandControls.CreateDefaultLabelFor("v_WindowState", this));
+            RenderedControls.Add(CommandControls.CreateDropdownFor("v_WindowState", this));
```
Comment (openbots-ab) [inline taskt.Studio/Commands/Window Commands/SetWindowStateCommand.cs:90]: You can use CreateDefaultDropDownGroupFor method here.

File: taskt.Studio/Commands/Window Commands/WaitForWindowToExistCommand.cs [hunk 15]
```diff
@@ -17,90 +18,71 @@ namespace taskt.Commands
...
-        public override void Refresh(IfrmCommandEditor editor)
-        {
-            base.Refresh();
-            WindowNameControl.AddWindowNames();
         }
+
         public override string GetDisplayValue()
         {
-            return base.GetDisplayValue() + " [Target Window: '" + v_WindowName + "', Wait Up To " + v_LengthToWait + " seconds]";
+            return base.GetDisplayValue() + $" [ Window '{v_WindowName}' - Timeout '{v_Timeout}']";
```
Comment (openbots-ab) [inline taskt.Studio/Commands/Window Commands/WaitForWindowToExistCommand.cs:85]:
```suggestion
            return base.GetDisplayValue() + $" [Window '{v_WindowName}' - Timeout '{v_Timeout}']";
```

**Review by openbots-ab — APPROVED** [review]

**Review by openbots-ab — CHANGES_REQUESTED** [review]
Please make a bunch of requested changes.

File: taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs (hunk 13 above)
Comment (openbots-ff) [reply taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs:65]: I think this command was intended to allow the user to set multiple engine preferences at runtime, but currently the only one available is the option to allow math to be done on variables. That's why using "To" would be too vague.

File: taskt.Studio/Commands/Window Commands/SetWindowStateCommand.cs (hunk 14 above)
Comment (openbots-ff) [reply taskt.Studio/Commands/Window Commands/SetWindowStateCommand.cs:90]: done

File: taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs (hunk 11 above)
Comment (openbots-ff) [reply taskt.Commands/System Commands/LaunchRemoteDesktopCommand.cs:25]: done

File: taskt.Commands/Data Commands/DateCalculationCommand.cs (hunk 1 above)
Comment (openbots-ff) [reply taskt.Commands/Data Commands/DateCalculationCommand.cs:213]: done

File: taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs (hunk 13 above)
Comment (openbots-ab) [reply taskt.Studio/Commands/Engine Commands/SetEnginePreferenceCommand.cs:66]: Then it's fine.

**Review by openbots-ab — APPROVED** [review]

**Pull request merged**